cpp/src/arrow/compute/kernels/scalar_if_else.cc

```diff
@@ -0,0 +1,285 @@
+                     const ArrayData& right, ArrayData* output) {
+  if (!cond.MayHaveNulls() && !left.MayHaveNulls() && !right.MayHaveNulls()) {
+    return Status::OK();  // no nulls to handle
+  }
+  const int64_t len = cond.length;
+
+  ARROW_ASSIGN_OR_RAISE(std::shared_ptr<Buffer> out_validity, ctx->AllocateBitmap(len));
+  arrow::internal::InvertBitmap(out_validity->data(), 0, len,
+                                out_validity->mutable_data(), 0);
+  if (right.MayHaveNulls()) {
+    // out_validity = right.val && ~cond.data
+    arrow::internal::BitmapAndNot(right.buffers[0]->data(), right.offset,
+                                  cond.buffers[1]->data(), cond.offset, len, 0,
+                                  out_validity->mutable_data());
+  }
+
+  if (left.MayHaveNulls()) {
+    // tmp_buf = left.val && cond.data
+    ARROW_ASSIGN_OR_RAISE(std::shared_ptr<Buffer> temp_buf,
+                          arrow::internal::BitmapAnd(
+                              ctx->memory_pool(), left.buffers[0]->data(), left.offset,
+                              cond.buffers[1]->data(), cond.offset, len, 0));
+    // out_validity = cond.data && left.val || ~cond.data && right.val
+    arrow::internal::BitmapOr(out_validity->data(), 0, temp_buf->data(), 0, len, 0,
+                              out_validity->mutable_data());
+  }
+
+  if (cond.MayHaveNulls()) {
+    // out_validity &= cond.val
+    ::arrow::internal::BitmapAnd(out_validity->data(), 0, cond.buffers[0]->data(),
+                                 cond.offset, len, 0, out_validity->mutable_data());
+  }
+
+  output->buffers[0] = std::move(out_validity);
+  output->GetNullCount();  // update null count
+  return Status::OK();
+}
+
+template <typename Type, bool swap = false, typename Enable = void>
+struct IfElseFunctor {};
+
+template <typename Type, bool swap>
+struct IfElseFunctor<Type, swap, enable_if_t<is_number_type<Type>::value>> {
+  using T = typename TypeTraits<Type>::CType;
+
+  static Status Call(KernelContext* ctx, const ArrayData& cond, const ArrayData& left,
+                     const ArrayData& right, ArrayData* out) {
+    ARROW_RETURN_NOT_OK(promote_nulls(ctx, cond, left, right, out));
+
+    ARROW_ASSIGN_OR_RAISE(std::shared_ptr<Buffer> out_buf,
+                          ctx->Allocate(cond.length * sizeof(T)));
+    T* out_values = reinterpret_cast<T*>(out_buf->mutable_data());
+
+    // copy right data to out_buff
+    const T* right_data = right.GetValues<T>(1);
+    std::memcpy(out_values, right_data, right.length * sizeof(T));
+
+    const auto* cond_data = cond.buffers[1]->data();  // this is a BoolArray
+    BitBlockCounter bit_counter(cond_data, cond.offset, cond.length);
+
+    // selectively copy values from left data
+    const T* left_data = left.GetValues<T>(1);
+    int64_t offset = cond.offset;
+
+    // todo this can be improved by intrinsics. ex: _mm*_mask_store_e* (vmovdqa*)
```

Comment:
I would assume memcpy already does this for you.

Reply:
Do you mean, load with mask?

Reply:
Sorry, I think I misunderstood the optimization you're thinking about. How would SIMD help here?

Reply:
Say, you first copy `right` to `ouput`. Then, `cond` becomes a mask to store `left` onto `output`. For that there are specialized SIMD instructions. 
https://software.intel.com/sites/landingpage/IntrinsicsGuide/#cats=Store&text=mask_store&expand=5564,5566

So, we can drop the `BitBlock` objects, and remove all the loops and memcpy inside the `while` loop. We'd have to handle the alignment though.